### DISCUSSION TITLE
All API Routes return "404 | Page not found" on Vercel

### DISCUSSION THREAD

**Discussion by sciffany** [vercel/next.js#44016] [Help]
**Problem**

I am working on a NextJS (monorepo) project with a login page and some API routes. All of them work well locally. But when I deploy them to Vercel, all of them return a standard 404 | Page not found page, instead of 200 OK.

Also note that all of these endpoints used to work in our previous deployments. Now, all of our previous deployments don't work anymore.

**Login Example**
For instance, when I sign in, I get redirected to 404, with the following error printed at the console:

```
GET https://domainname.com/api/auth/error 404
GET https://domainname.com/api/auth/session 404
[next-auth][error][CLIENT_FETCH_ERROR] 
https://next-auth.js.org/errors#client_fetch_error Unexpected token '<', "<!DOCTYPE "... is not valid JSON
POST https://domainname.com/api/auth/_log 404 
```
This is despite having a proper authentication endpoint at `pages/api/auth/[...nextauth].ts`

```
import NextAuth from 'next-auth'
import { type NextAuthOptions } from 'next-auth'
import EmailProvider from 'next-auth/providers/email'

import { prisma } from '@onlyhelper/db'
import { PrismaAdapter } from '@next-auth/prisma-adapter'

export const authOptions: NextAuthOptions = {
  // Configure one or more authentication providers
  adapter: PrismaAdapter(prisma),
  providers: [
    EmailProvider({
      server: process.env.ALL_EMAIL_SERVER,
      from: 'hello@domainname.com'
    })
  ],
  secret: 'secret',
  session: {},
  callbacks: {
    session({ session, user }) {
      session.user.id = user.id
      return session
    }
  }
}

export default NextAuth(authOptions)
```

**API Example**
When I call my endpoint `domainname.com/api/checkout`, I also get a 404 page on Insomnia. 

This is despite having an endpoint at `/pages/api/checkout.ts` which has the following code.

```
export default async function checkout(
  request: NextApiRequest,
  response: NextApiResponse
) {
// database calls that don't ever return 404
return res.status(200).status({ message: "success" })
}
```

**What I've tried**
- Pointed local instance to production database, and logged in successfully, to confirm this isn't a database configuration/corrupted data issue
- Used ngrok to expose and test API endpoints locally. It works! So it's only the Vercel deployment that's returning 404 Page
- Researched Google and Stack Overflow for similar issues.
- Removing other folders named "api" just to ensure that Vercel doesn't get confused with folder names.
- Checked Vercel "Source" page to see that all endpoints are present 

**Question**
- What could be causing this?
- Is is possible that our serverless function limit for the month has been reached?

Thanks in advance for all your help!

**Comment by icyJoseph** [comment]
First thing first, are you using `next export`?

**Reply by sciffany** [reply]
No we're not using `next export` for this folder in the monorepo. We do have another folder using `next export`, but it's not deployed to Vercel.

**Reply by sciffany** [reply]
I've deleted the unused folder which uses `next export` for simplicity. Still returning 404.

**Reply by icyJoseph** [reply]
Strange then. If you had hit your limits you'd get a 5xx I think. Are you also getting 404 on pages? Not just API routes?

**Reply by sciffany** [reply]
Just API routes, pages are fine.

**Reply by icyJoseph** [reply]
Could you try to drop some logs inside the API route handlers, see if they get triggered at least.

Do you see any information in the Vercel dashboard? Each deployment has something called function logs, there should be more clues there.

**Reply by sciffany** [reply]
That's a great idea. Will try it out and let you know. Thanks!

**Reply by sciffany** [reply]
Function logs aren't displaying my console.logs. But I did find that when I return 403, 403 is returned instead of 404.

**Reply by icyJoseph** [reply]
Do you have any logic to check the `method` on the request? Perhaps it expects POST/GET etc?

Do you have any rewrite/redirect logic that might be interfering with API routes?

**Reply by sciffany** [reply]
No, no method checking logic.
No, no rewrite/redirect logic.

**Reply by sciffany** [reply]
Seems like the API could be returning 404 because of varied reasons. The newer API routes are not handling database requests properly, and for some reason, are returning 404 instead of 500. I still have yet to figure out why the login API routes are returning 404.

**Reply by icyJoseph** [reply]
Mmm I'd ask the Support channel on the Vercel site, this looks quite strange. I'd only guess that somehow a CDN has cached your api path, but that sounds farfetched.

**Reply by sciffany** [reply]
Alright, will look into that. Thanks for all your help, @icyJoseph! Really appreciate it.
